### PR TITLE
Added Cepik integration for trainee registration

### PR DESCRIPTION
**Pull request by maonat**
Mock Cepik integration for trainee registration

## Description
- Upgraded trainee object to allow containing new required dataOfBirth property
- Integrated CepikService to allow returning mocked randomized values (firstName, lastName, pesel)
- Created new signup endpoint to allow users to register by providing dateOfBirth and PKK number
- Allowed pesel to be nullable in case of non-polish applicants
- Added new dateOfBirth field in FE and nullable pesel as it's not required and returned by cepik
## What part of the app is affected?
* [x] Frontend
* [x] Backend
* [ ] Infrastructure

### CODE REVIEW

**Comment by maonat** [comment]
@bibixx Changed FE as well because had pesel non-nullable. Added as well new field for dateOfBirth.

File: backend/src/cepik/cepik.service.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
Comment (bibixx) [inline backend/src/cepik/cepik.service.ts]: You shouldn't need to disable this rule.
Also for the future, please try to use `// eslint-disable-next-line` rule, to limit the amount of affected lines

File: backend/src/cepik/cepik.service.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+@Injectable()
+export class CepikService {
+  registerUser(
+    pkkNumber: string,
+    dateOfBirth: Date,
+  ): Try<CepikResponseData, 'CEPIK_ERROR'> {
+    // Adding testing CEPIK_ERROR for testing purposes whenever
+    // someone creates a new user with dateOfBirth equal to
+    // 1st of January
+    if (dateOfBirth.getDay() === 1 && dateOfBirth.getMonth() === 0) {
```
Comment (bibixx) [inline backend/src/cepik/cepik.service.ts]: It should be `dateOfBirth.getDate() === 1` instead
> The getDay() method returns the day of the week for the specified date according to local time, where 0 represents Sunday.
> -https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getDay

File: backend/src/trainees/trainees.controller.ts [hunk 3]
```diff
@@ -59,16 +66,34 @@ export class TraineesController {
   @ApiResponse({
     type: TraineeAddNewResponseDto,
   })
-  @ApiBody({ type: TraineeAddNewRequestDto })
-  @Post()
-  async create(
-    @Body() { trainee }: TraineeAddNewRequestDto,
-    @Request() req: ExpressRequest,
+  @ApiBody({ type: TraineeAddNewRequestSignupDto })
+  @Post('/signup')
```
Comment (bibixx) [inline backend/src/trainees/trainees.controller.ts:72]: This endpoint should be available to the public, shouldn't it?

File: backend/src/trainees/trainees.controller.ts [hunk 4]
```diff
@@ -85,6 +110,49 @@ export class TraineesController {
...
+  ): Promise<TraineeAddNewResponseDto> {
+    const isHttps = req.protocol === 'https';
+    const createTraineeCall = await this.traineesService.create(trainee);
+
+    if (!createTraineeCall.ok) {
+      const { error } = createTraineeCall;
+
+      if (error === 'TRAINEE_OR_USER_FOUND') {
+        throw new ConflictException(
+          'There is already a trainee which has the same pesel or an user with the same email',
```
Comment (bibixx) [inline backend/src/trainees/trainees.controller.ts]: Quite an interesting grammar fact 😄 
https://grammarhow.com/a-user-or-an-user/

```suggestion
          'There is already a trainee which has the same pesel or a user with the same email',
```

File: shared/src/dto/trainee/trainee.dto.ts [hunk 5]
```diff
@@ -55,12 +94,19 @@ export class DtoUpdateTrainee {
...
+  @Length(11, 11)
+  pesel?: string | null;
 
   @ApiProperty()
   @IsOptional()
   pkk?: string;
 
+  @ApiProperty()
+  @IsNotEmpty()
+  @Type(() => Date)
```
Comment (bibixx) [inline shared/src/dto/trainee/trainee.dto.ts]: Proper validation for this being a date is missing. Please take a look at how I did it in the `availability.dto.ts` file.
Also take a look at how I have used the `@ApiProperty()` decorator

File: shared/src/dto/user/user.dto.ts [hunk 6]
```diff
@@ -78,6 +78,23 @@ export class DtoCreateUser {
...
+export class DtoCreateUserSignup {
+  @ApiProperty()
+  @IsNotEmpty()
+  @IsEmail()
+  email: string;
+
+  @ApiProperty()
+  @IsNotEmpty()
+  @IsBoolean()
+  isActive: boolean;
```
Comment (bibixx) [inline shared/src/dto/user/user.dto.ts]: Well I don't think the user should be able to specify whether they are active or not 🙂
Also isn't this DTO missing a password property?

**Review by bibixx — CHANGES_REQUESTED** [review]

File: backend/src/trainees/trainees.controller.ts (hunk 3 above)
Comment (maonat) [reply backend/src/trainees/trainees.controller.ts:72]: I'm not sure I understand what you are saying right now. Could you please rephrase it @bibixx ?

Comment (bibixx) [reply backend/src/trainees/trainees.controller.ts:72]: Well, right now you have to be logged in as an Admin or Instructor to access this endpoint

File: shared/src/dto/user/user.dto.ts (hunk 6 above)
Comment (maonat) [reply shared/src/dto/user/user.dto.ts]: This was made by someone else before, probably didn't think about fixing the missing thing 🙈

File: backend/src/trainees/trainees.controller.ts (hunk 3 above)
Comment (maonat) [reply backend/src/trainees/trainees.controller.ts:72]: @bibixx Please double check if the changes are ok right now.

File: shared/src/dto/trainee/trainee.dto.ts [hunk 7]
```diff
@@ -32,14 +45,49 @@ export class DtoCreateTrainee {
...
   @IsNotEmpty()
   pkk: string;
 
+  @ApiProperty({
+    type: 'string',
+    format: 'YYYY-mm-DDTHH:mm:ss.SZ',
+  })
+  @IsNotEmpty()
+  @Type(() => Date)
+  dateOfBirth: ApiDate;
```
Comment (bibixx) [inline shared/src/dto/trainee/trainee.dto.ts:92]: It still isn't being validated as Date. It's missing `@IsDate`

File: backend/src/cepik/cepik.service.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import faker from '@faker-js/faker';
```
Comment (bibixx) [inline backend/src/cepik/cepik.service.ts]: One more note — let's use this import to get polish names 😉 
```ts
import * as faker from '@faker-js/faker/locale/pl';
```

File: shared/src/dto/user/user.dto.ts [hunk 9]
```diff
@@ -96,6 +93,11 @@ export class DtoCreateUserSignup {
   @IsEmail()
   email: string;
 
+  @ApiProperty()
+  @IsNotEmpty()
+  @IsString()
+  password: string | null;
```
Comment (bibixx) [inline shared/src/dto/user/user.dto.ts]: Length validation is missing

File: shared/src/dto/user/user.dto.ts [hunk 10]
```diff
@@ -78,6 +87,24 @@ export class DtoCreateUser {
...
+  @ApiProperty()
+  @IsNotEmpty()
+  @IsEmail()
+  email: string;
+
+  @ApiProperty()
+  @IsNotEmpty()
+  @IsString()
+  @Length(8, 64)
+  password: string | null;
```
Comment (bibixx) [inline shared/src/dto/user/user.dto.ts]: Although in TS we have specified that this field can be `null`, actually the validation doesn't allow it to be null

**Review by bibixx — APPROVED** [review]

**Pull request merged**